src/Apps/Legion/Components/Diagram/Diagram.tsx

```diff
@@ -0,0 +1,137 @@
+            if (
+                parentRef.current.clientWidth /
+                    imageOriginalDimensionRef.current.width >
+                parentRef.current.clientHeight /
+                    imageOriginalDimensionRef.current.height
+            ) {
+                setImgHeight('100%');
+                setImgWidth('fit-content');
+            } else {
+                setImgWidth('100%');
+                setImgHeight('fit-content');
+            }
+
+            setAnnotationPlacements(
+                annotations.reduce((acc, a) => {
+                    if (
+                        !a.values[PID_EXTRACTED_PROPERTIES.Confidence] ||
+                        a.values[PID_EXTRACTED_PROPERTIES.Confidence] >=
+                            CONFIDENCE_TRESHOLD
+                    ) {
+                        acc.push({
+                            ...a,
+                            left: `calc(${
+                                (a.values[PID_EXTRACTED_PROPERTIES.X] /
+                                    imageOriginalDimensionRef.current.width) *
+                                100
+                            }% - ${ANNOTATION_X_PADDING}px)`,
+                            top: `${
+                                (a.values[PID_EXTRACTED_PROPERTIES.Y] /
+                                    imageOriginalDimensionRef.current.height) *
+                                100
+                            }%`
+                        });
+                    }
+                    return acc;
+                }, [])
+            );
+            setIsReady(true);
+        }
+    }, [annotations, isImageLoaded, parentRef]);
+
+    // callbacks
+    const handleOnImgLoad = useCallback(({ target: image }) => {
+        imageOriginalDimensionRef.current = {
+            width: image.naturalWidth,
+            height: image.naturalHeight
+        };
+        setIsImageLoaded(true);
+    }, []);
+
+    // styles
+    const classNames = useClassNames();
+
+    logDebugConsole('debug', 'Render');
+
+    return (
+        <div
+            className={classNames.root}
+            style={{
+                width: !isReady ? '100%' : imgWidth,
+                height: !isReady ? '100%' : imgHeight
+            }}
+        >
+            {!isReady && (
+                <Spinner
+                    className={classNames.spinnerWrapper}
+                    size={SpinnerSize.medium}
+                    label={t('loading')}
+                />
+            )}
+            <Image
+                ref={imageRef}
+                width={imgWidth}
+                height={imgHeight}
+                onLoad={handleOnImgLoad}
+                src={imageUrl}
+            />
+            {annotationPlacements.map((a, idx) => (
+                <div
+                    key={imageUrl + idx}
```

Comment:
what do we gain by having hte image url in the key? Doesn't every item in the map have the same url? Can we pick something more unique to each item so that the key is more guaranteed to be unique?
Also, we have seen issues in the past with using index as the primary discriminator in keys so let's not rely on that here since it causes issues if the ordering of items changes in the collection. React then thinks nothing changed and won't rerender those items where the key still matches even though the contents of that item should have been updated.

Reply:
the idea was in case of multiple diagrams are shown up in the same UI, the image url could be used as unique key that I do not see any problem with unless used along with an idx. So, I can generate an id with useId hook instead of the image url.

Reply:
Gotcha. Actually, instead of useId I would just use something unique (or a combo) from the object. The idea of key is that it is generated based off of the item being rendered so it knows whether to rerender the child or not. If the key doesn't change, then when the parent renders it will use the existing value in the virtual DOM instead of recreating fresh DOM elements.
So in this case, if you can use icon+text or something that should be enough.